packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -956,6 +987,10 @@ class RenderParagraph extends RenderBox
         newChildren.add(newChild);
       }
     }
+    // Makes sure we exhausted the list.
+    assert(childIndex == children.length);
```

Comment:
There seem valid cases where childIndex < children.length.

In my test where this used to crash, I get childIndex: 2, children.length 7, placeholderIndex: 8 at this point, as a reference.

Reply:
That is unexpected, we have to annotate all the semantics children, or they may not have the correct semantics. Do you have a repro example?

Reply:
Does my complicated example help? That is pretty close to what customer:money is actually doing. https://github.com/flutter/flutter/issues/69787#issuecomment-722169455

Reply:
We may be missing the use case where a WidgetSpan is bubbling up more than one semantic node, something like:

```dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Directionality(
      textDirection: TextDirection.ltr,
      child: Center(
        child: Container(
          color: Colors.green,
          child: RichText(
            text: TextSpan(
              children: [
                TextSpan(
                  text: 'Start\n',
                  recognizer: TapGestureRecognizer()..onTap = () { },
                ),
                WidgetSpan(
                  child: Icon(
                    Icons.edit,
                    size: 16,
                  ),
                ),
                WidgetSpan(
                  child: Row(
                    children: [
                      Semantics(
                        container: true,
                        child: Icon(
                          Icons.edit,
                          size: 16,
                          semanticLabel: 'foo',
                        ),
                      ),
                      Semantics(
                        container: true,
                        child: Icon(
                          Icons.edit,
                          size: 16,
                          semanticLabel: 'bar',
                        ),
                      ),
                    ],
                  ),
                ),
                TextSpan(text: 'End'),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
```

@chunhtai can you try this with your change and see if it works? There can probably be more than one semantics node in `children` for a given `placeholderIndex`.

Reply:
yes, this pr missed this use case, I will update it to make sure it works

Reply:
Also feel free to ping me on chat if would be helpful, happy to walk through it or help dig. I don't have a simple setup, but have a snapshot with the above crash.